src/Microsoft.AspNet.Mvc.Core/ResponseCacheAttribute.cs

```diff
@@ -0,0 +1,132 @@
+        public bool NoStore { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value for the Vary response header.
+        /// </summary>
+        public string VaryByHeader { get; set; }
+
+        // <inheritdoc />
+        public override void OnActionExecuting([NotNull] ActionExecutingContext context)
+        {
+            // If there are more filters which can override the values written by this filter,
+            // then skip execution of this filter.
+            if (ContainsInnerFilter(context))
+            {
+                return;
+            }
+
+            var headers = context.HttpContext.Response.Headers;
+
+            // Clear all headers
+            headers.Remove("Vary");
+            headers.Remove("Cache-control");
+            headers.Remove("Pragma");
+
+            // Set appropriate headers
+            if (_duration == null)
+            {
+                throw new InvalidOperationException(Resources.ResponseCache_SpecifyDuration);
+            }
+
+            if (!string.IsNullOrEmpty(VaryByHeader))
+            {
+                headers.Set("Vary", VaryByHeader);
+            }
+
+            if (NoStore)
+            {
+                headers.Set("Cache-control", "no-store");
+                // Cache-control: no-store, no-cache is valid.
+                if (Location == ResponseCacheLocation.None)
+                {
+                    headers.Append("Cache-control", "no-cache");
+                    headers.Set("Pragma", "no-cache");
+                }
+            }
+            else
+            {
+                string cacheControlValue = null;
+                switch (Location)
+                {
+                    case ResponseCacheLocation.Any:
+                        cacheControlValue = "public";
+                        break;
+                    case ResponseCacheLocation.Client:
+                        cacheControlValue = "private";
+                        break;
+                    case ResponseCacheLocation.None:
+                        cacheControlValue = "no-cache";
+                        headers.Set("Pragma", "no-cache");
+                        break;
+                }
+
+                cacheControlValue = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}{1}max-age={2}",
+                    cacheControlValue,
+                    cacheControlValue != null? "," : null,
+                    Duration);
+
+                if (cacheControlValue != null)
+                {
+                    headers.Set("Cache-control", cacheControlValue);
+                }
+            }
+        }
+
+        // internal for Unit Testing purposes.
+        internal bool ContainsInnerFilter([NotNull] ActionExecutingContext context)
+        {
+            // Return true if there are any inner filters of type ResponseCacheAttribute.
+            return context.Filters.OfType<ResponseCacheAttribute>().Any(item => item.Order < Order);
```

Comment:
We should create an interface instead. Create a marker interface called `IResponseCacheFilter` and then look for that.

Also, the check here: `.Any(item => item.Order < Order);` is not correct. The system does ordering primarily by scope, think of `.Order` as an override. You need to walk the list and look for `this` and then see who's after.